### ISSUE TITLE
Question about  movavg_cls_threshold in training nuscenes

### ISSUE THREAD

**Issue by Journeyfu** [mercedes-benz/selfsupervised_flow#8]
Thanks for your good project to the self-supervised scene flow.
But I have a question about movavg_cls_threshold used in https://github.com/mercedes-benz/selfsupervised_flow/blob/master/unsup_flow/experiment/raft/raft_model.py#L32. I noticed that the config of nuscenes experiment (https://github.com/mercedes-benz/selfsupervised_flow/blob/master/unsup_flow/config/config.yml#L209) explicit set "num_still_points" parameter, which means the update process of movavg_cls_threshold will introduce the prior of moving_mask and the proportion of dynamic and static points, instead of a completely self supervised training pipeline.
Is this a wrong in nuscenes config?

**Comment by baurst** [comment]
Thank you for your interest in our project. We will investigate in the upcoming week, please give us some time and we will get back to you. Thank you!

**Comment by demmerichs** [comment]
Hi @Journeyfu, thanks again for your kind words and making us aware of this issue.

**TLDR**: We made a mistake there, we fixed it (see above linked PR), reran the experiments and found no significant changes in performance (very slightly worse but probably not significant w.r.t. noise between different trainings) and especially it does not change any of the assessments made in the paper. Read on for more details and concrete score values from our rerun experiments.

We want to share our thoughts on how we think this bug was introduced and why, and what we did and will do in the future to avoid this happening again:
1. *Assertions*: As you can see, we made extensive use of assertions in order to catch bugs, wrong usage of parameter combinations and conflicting modes, as well as improper values and dimensions of tensors handed between different modules as early and with as high of a coverage as we can get as individuals working on this. A concrete example related to the origin of the bug would be assertions inside the [MovingAverageThreshold](https://github.com/mercedes-benz/selfsupervised_flow/blob/652b54475e66cc7b8837d7c59641f60688e2053a/unsup_flow/experiment/movavg_cls_threshold.py#L11-L174) class, making sure that when `num_still` is set to `None` that this should be equivalent to not passing `moving_mask` as input (e.g. default `None`) to the `update` routine. This effectively made `num_still` a flag for using this class in an unsupervised or supervised manner.
2. *Self-documenting code*: We see still space for improvement regarding this aspect, but we did our best to write code quite verbose, especially in higher levels where a lot of components come together and interact. With the pace at which research is happening in our field, and the small number of individuals working on a project, an extensive documentation is a waste of time and is therefor often omitted by others, and us as well. Sadly, the above-mentioned `num_still` flag is a negative example of this. Of course the name itself does not indicate in any way, that it is also used as flag for a supervised mode, in contrast to a non-supervised one. We added a few lines of code, that better self-document this behavior in the commit 308643c01a7d2c85501128807ae11ba74d3ebf77. With this modification, our code would have thrown an assertion when executed with the wrong configuration of `num_still` set in an unsupervised training.
3. *Avoiding time-pressured hacks*: With the pace of our research community it is important to develop new methods fast, in order to stay relevant on conferences. Sadly, this puts a lot of pressure on the code quality and removes some of the sanity checks, refactoring and testing, required to spot bugs as the one @Journeyfu found. We fixed this now by setting the value for `num_still` to `None` in case of an unsupervised training (06d1f658d3147748377a17b5dd51e474c546992c). Sadly, this bug was able to survive in our code base even though we had measures discussed above in place. We cannot recall, more than a year after writing the paper, but we suspect that during development, the example assertion mentioned in point 1, was one of the last-minute-assertions we had to fix in order to run our experiments for the ICCV deadline. The assertion must have said something like `either set num_still to None or provided the moving_mask`. Under the time-pressure, we must have lost the context information, that this was an assertion from an unsupervised training, and therefor the correct we to handle this assertion would have been, to set `num_still` to `None`. However, based on the code we found [here](https://github.com/mercedes-benz/selfsupervised_flow/blob/652b54475e66cc7b8837d7c59641f60688e2053a/unsup_flow/experiment/losses.py#L711-L739), we assume that we did the opposite, and instead decided to provide the `moving_mask`, overseeing the fact that this was supposed to be an unsupervised training. We jumped directly to the line of the calling function, created the `moving_mask` on the fly, and restarted the training. If we would have time to check and look a little bit closer, we could have noticed, that this code change happened inside the `UnsupervisedLoss` class (definition line was sadly way up outside our field of view), directly contradicting the supposed thing this class was supposed to do. Sadly, the *self-documenting* code did not help avoid this bug here. We removed the construction of the `moving_mask` from this class in 5737d313e586a855c604c4922d9a17d9c64b702c.

With the above improvements and fixes implemented, we redid the experiment training and evaluating on NuScenes in an unsupervised manner. Although we provided in the original train runs quite a bit of information by introducing a pointwise moving mask, we already suspected that this might be not a big game changer for the following reasons:
1. It influences only a single scalar directly, the selection of the threshold. The training is only indirectly influenced by this, using the threshold to create the dynamic and static mask for flow aggregation and separation.
2. The moving mask was just used as weights, and no part of the network was directly supervised with this information.
3. What was weighted, was the KNN distance based on aggregated vs. non-aggregated flow. This KNN signal was extremely noisy (we mentioned this in the supplementary material) and was not a really strong signal to begin with.Therefor the quality of the threshold is probably quite low, even with the usage of weights.

Nonetheless, it is important to check our assumptions.
For now, we just repeated the best-performing unsupervised NuScenes training, which is row 2 in the ablation study table 7.
| Table 7, row 2 | Mov. AEE  | Stat. AEE | 50-50 AEE |
| ------------- | ------------- | ------------- | ------------- |
| original values from paper  | **0.1050**  | 0.0925 | **0.0987** |
| values with bugfixes from above PR | 0.1116  | **0.0870**  | 0.0993 |

As one might expect, this shifts the performance focus slightly to the static points, but the change is not big and the KNN signal not clear enough to produce extreme overfitting to one or the other.

**Comment by Journeyfu** [comment]
The performance difference is unexpectedly small. Thank you again for solid ablation experiments!
